packages/data/src/components/use-select/index.js

```diff
@@ -104,7 +105,100 @@ const renderQueue = createQueue();
  * ```
  * @return {UseSelectReturn<T>} A custom react hook.
  */
+function Selecter( registry ) {
+	const NOTHING = {};
+	let lastMapSelect = NOTHING;
+	let lastMapResult = NOTHING;
+	let lastMapResultValid = false;
+
+	const NULLSCRIBER = () => {
+		lastMapResultValid = false;
+		return () => {};
+	};
+	let subscribe = NULLSCRIBER;
+
+	function createSubscriber( stores ) {
+		if ( stores.length === 0 ) {
+			return NULLSCRIBER;
+		}
+
+		return ( lis ) => {
+			lastMapResultValid = false;
+			const unsubs = stores.map( ( storeName ) => {
+				return registry.subscribe( () => {
+					lastMapResultValid = false;
+					lis();
+				}, storeName );
+			} );
+
+			return () => {
+				// The return value of the subscribe function could be undefined if the store is a custom generic store.
+				unsubs.forEach( ( unsub ) => unsub?.() );
+			};
+		};
+	}
+
+	function select( mapSelect, resub ) {
+		function getValue() {
+			if ( lastMapResultValid && mapSelect === lastMapSelect ) {
+				return lastMapResult;
+			}
+
+			const mapResult = mapSelect( registry.select, registry );
+
+			if (
+				lastMapResult === NOTHING ||
+				! isShallowEqual( lastMapResult, mapResult )
+			) {
+				lastMapResult = mapResult;
+			}
+			lastMapResultValid = true;
+
+			return lastMapResult;
+		}
+
+		if ( lastMapResultValid && mapSelect === lastMapSelect ) {
+			return { getValue, subscribe };
+		}
+
+		const listeningStores = { current: null };
```

Comment:
Is there a reason why we're not defaulting to an empty array but to a `null` here?

Reply:
Nobody would use the empty array. The `__unstableMarkListeningStores` function always creates a new array and assigns it to the ref.